apps/chat/conversation.py

```diff
@@ -208,9 +209,18 @@ def compress_chat_history_from_messages(
     input_message_tokens = llm.get_num_tokens_from_messages(input_messages)
     history_tokens = llm.get_num_tokens_from_messages(history)
     first_pass_done = False  # ensure we do at least one loop
-    while not first_pass_done or (history and history_tokens + summary_tokens + input_message_tokens > max_token_limit):
+
+    while (
+        not first_pass_done
+        # Check `_tokens_exceeds_limit` here, since the generated summary can cause a token limit breach
+        or _tokens_exceeds_limit(
+            history, token_count=(history_tokens + summary_tokens + input_message_tokens), limit=max_token_limit
+        )
+    ):
         first_pass_done = True
-        while history and history_tokens + summary_tokens + input_message_tokens > max_token_limit:
+        while _tokens_exceeds_limit(
+            history, token_count=(history_tokens + summary_tokens + input_message_tokens), limit=max_token_limit
+        ) or _messages_exceeds_limit(history, input_messages):
             pruned_memory.append(history.pop(0))
```

Comment:
Instead of just popping off one message we could trim it to the correct size in one go which would reduce the number of loops.

Reply:
Good idea, I'll open a new PR for this

Reply:
https://github.com/dimagi/open-chat-studio/pull/1243